src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouter.cs

```diff
@@ -0,0 +1,841 @@
+    /// This class connects IPC Server - TCP Server router instances.
+    /// Supports NamedPipes/UnixDomainSocket server and TCP/IP server.
+    /// </summary>
+    internal class IpcServerTcpServerRouter : TcpServerRouter, IIpcServerTransportCallbackInternal
+    {
+        readonly string _ipcServerPath;
+
+        IpcServerTransport _ipcServer;
+
+        public int IpcServerConnectTimeout { get; set; } = Timeout.Infinite;
+
+        public IpcServerTcpServerRouter(string ipcServer, string tcpServer, int runtimeTimeoutMs, ILogger logger)
+            : base(tcpServer, runtimeTimeoutMs, logger)
+        {
+            _ipcServerPath = ipcServer;
+            if (string.IsNullOrEmpty(_ipcServerPath))
+                _ipcServerPath = GetDefaultIpcServerPath();
+
+            _ipcServer = IpcServerTransport.Create(_ipcServerPath, IpcServerTransport.MaxAllowedConnections, false);
+            _tcpServer.TransportCallback = this;
+        }
+
+        public static string GetDefaultIpcServerPath()
+        {
+            int processId = Process.GetCurrentProcess().Id;
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}");
+            }
+            else
+            {
+                DateTime unixEpoch;
+#if NETCOREAPP2_1_OR_GREATER
+                unixEpoch = DateTime.UnixEpoch;
+#else
+                unixEpoch = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
+#endif
+                TimeSpan diff = Process.GetCurrentProcess().StartTime.ToUniversalTime() - unixEpoch;
+                return Path.Combine(PidIpcEndpoint.IpcRootPath, $"dotnet-diagnostic-{processId}-{(long)diff.TotalSeconds}-socket");
+            }
+        }
+
+        public override void Start()
+        {
+            base.Start();
+            _logger.LogInformation($"Starting IPC server ({_ipcServerPath}) <--> TCP server ({_tcpServerAddress}) router.");
+        }
+
+        public override Task Stop()
+        {
+            _ipcServer?.Dispose();
+            return base.Stop();
+        }
+
+        public override async Task<ConnectedRouter> ConnectRouterAsync(CancellationToken token)
+        {
+            Stream tcpServerStream = null;
+            Stream ipcServerStream = null;
+
+            Logger.LogDebug($"Trying to connect new router instance.");
+
+            try
+            {
+                using CancellationTokenSource cancelConnectRouter = CancellationTokenSource.CreateLinkedTokenSource(token);
+
+                // Connect new tcp server endpoint.
+                using var tcpServerStreamTask = ConnectTcpStreamAsync(cancelConnectRouter.Token);
+
+                // Connect new ipc server endpoint.
+                using var ipcServerStreamTask = ConnectIpcStreamAsync(cancelConnectRouter.Token);
+
+                await Task.WhenAny(ipcServerStreamTask, tcpServerStreamTask).ConfigureAwait(false);
```

Comment:
I think the naming scheme is too uniform and makes it a little confusing which actor is doing the connecting and which actor is doing the listening. My understanding of the intent is that the call to `ConnectTcpStreamAsync` is starting a TCP server and _listening_ on that server for a new connection from a runtime, while the call to `ConnectIpcStreamAsync` is connecting to a local tool's IPC server. Would it be possible to either change `ConnectTcpStreamAsync` to `AcceptTcpStreamAsync`, or not use `connect` in the naming to remove the association with the `listen+accept/connect` API naming scheme?

Reply:
I will try to clarify this.